File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
def solution(info, edges):
    def dfs(sheep, wolf):
        if sheep <= wolf:
            return
        
        answer.append(sheep)
        
        for parent, child in edges:
            check = (info[child] == 1)
            
            if not visited[parent] or visited[child]:
                continue
            
            visited[child] = True
            dfs(sheep + (not check), wolf + check)
            visited[child] = False
    
    
    answer = []
    visited = [False] * len(info)
    visited[0] = True
    
    dfs(1, 0)
    
    return max(answer)

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {

    private int[] info;
    private int[][] edges;
    private int answer;

    public int solution(int[] info, int[][] edges) {
        this.info = info;
        this.edges = edges;
        this.answer = 0;

        boolean[] visited = new boolean[info.length];
        visited[0] = true;

        dfs(1, 0, visited);
        return answer;
    }

    private void dfs(int sheep, int wolf, boolean[] visited) {
        if (sheep <= wolf) return;

        answer = Math.max(answer, sheep);

        for (int[] edge : edges) {
            int parent = edge[0];
            int child = edge[1];

            if (!visited[parent] || visited[child]) continue;

            visited[child] = true;
            if (info[child] == 1) dfs(sheep, wolf + 1, visited);
            else dfs(sheep + 1, wolf, visited);
            visited[child] = false;
        }
    }
}